### PR TITLE
[Runtime] Search through protocols to resolve protocol declarations.

### PR DESCRIPTION
**Pull request by DougGregor**
Introduce a new section containing protocol descriptor references, and use this section to allow us to find protocol descriptors by mangled name. Use this functionality to fully support protocol composition types within `_typeForMangledName`.

### CODE REVIEW

**Comment by DougGregor** [comment]
@jckarter you might be interested in this

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: test/Runtime/demangleToMetadata.swift [hunk 1]
```diff
@@ -78,12 +78,23 @@ DemangleToMetadataTests.test("metatype types") {
...
+func f1_composition(_: P1 & P2) { }
+func f1_composition_superclass(_: C & P1 & P2) { }
+
 DemangleToMetadataTests.test("existential types") {
   // Any, AnyObject
   expectEqual(type(of: f2_any_anyobject), _typeByMangledName("yyyp_yXltc")!)
 
-  // FIXME: References to protocols.
-  // FIXME: References to superclass.
+  // References to protocols.
```
Comment (slavapestov) [inline test/Runtime/demangleToMetadata.swift]: Just for completeness, can you test `P & AnyObject` where P is a non-class-bound protocol?

Comment (DougGregor) [reply test/Runtime/demangleToMetadata.swift]: Will do!

**Comment by DougGregor** [comment]
Must have cargo-culted ELF handling badly

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Pull request merged**